### DISCUSSION TITLE
Mac Arthur HAT compatilibity

### DISCUSSION THREAD

**Discussion by den200** [bareboat-necessities/lysmarine_gen#390] [Q&A] [answered]
Hey there, 

Does anyone know if BBN OS is compatible with  the new Mac Arthur HAT ?

**Comment by mgrouch** [comment]
Latest BBN OS is Bookworm Raspberry based so you should check with your hardware vendor how they support at least that. Or if they have BBN OS specific instructions.

**Comment by mgrouch** [comment]
Pi settings (untested) for NMEA 2000 but something along these lines:

````
/boot/config.txt
...
#dtoverlay=disable-bt
#dtoverlay=w1-gpio
enable_uart=1 
dtparam=i2c_arm=on 
dtparam=spi=on
dtoverlay=mcp251xfd,oscillator=20000000,interrupt=25,spi0-1
#dtoverlay=spi-bcm2835
...


/etc/network/interfaces.d/can0
...
# physical can interfaces
allow-hotplug can0
  iface can0 can static
  bitrate 250000
  pre-up ip link set can0 type can restart-ms 100
  down /sbin/ip link set $IFACE down
  up /sbin/ifconfig $IFACE txqueuelen 10000
...

  root@roam:~# ifconfig
can0: flags=193<UP,RUNNING,NOARP>  mtu 16
        unspec 00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00  txqueuelen 10  (UNSPEC)
        RX packets 3920303  bytes 31362424 (29.9 MiB)
        RX errors 0  dropped 623  overruns 0  frame 0
        TX packets 4  bytes 12 (12.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
````

**Comment by jsextl** [comment]
Works perfect without problems - only difference is, that BBN has a separate service for the CAN interface, so you don't have to configure /etc/network/interfaces manually.

If anybody needs detailed instructions, just let me know and I can copy&paste exact settings / configuration.

Best regards,
Jonas

**Reply by mgrouch** [reply]
Would be nice if you could post here your exact working config. 

Thanks

**Comment by mgrouch** [comment]
> Works perfect without problems - only difference is, that BBN has a separate service for the CAN interface, so you don't have to configure /etc/network/interfaces manually.
> 
> If anybody needs detailed instructions, just let me know and I can copy&paste exact settings / configuration.
> 
> Best regards, Jonas

Would be nice if you could post here your exact working config. And also instructions regarding service. 

Thanks

**Reply by jsextl** [reply, marked as answer]
Sure no problem.

1. Enable driver in /boot/firmware/config.txt:
```
[...]
# enable CANBUS
dtoverlay=mcp251xfd,oscillator=20000000,interrupt=25,spi0-1
[...]
```

I put this configuration after the last "dtoverlay" statement present in the file of a fresh default installation.

2. Reboot

3. Enable can bus interface

```
sudo systemctl enable socketcan-interface
sudo systemctl start socketcan-interface
```

If everything is properly setup on the hardware side, you should receive the CANBUS traffic via:

`candump can0`

**Comment by mgrouch** [comment]
Relevant to the discussion (non isolated NMEA 2000)

https://github.com/OpenMarine/MacArthur-HAT/issues/14


An article about ground loops and isolation 

https://actisense.com/news/why-nmea-isolation-is-critical-for-your-network/

**Comment by mgrouch** [comment]
UARTs of the MacArthur hat for NMEA 0183 on BBN Marine OS

Raspberry Pi 5

NMEA 1 in/out: UART2
NMEA 2 in/out: UART4

in /boot/firmware/config.txt

````
dtparam=uart0=on
dtoverlay=uart2-pi5
dtoverlay=uart4-pi5
````

Raspberry Pi 4 and 3

NMEA 1 in/out: UART3
NMEA 2 in/out: UART5 

````
dtoverlay=disable-bt
dtoverlay=uart3
dtoverlay=uart5
````

Reboot after editing /boot/firmware/config.txt

and check devices created

````
ls  -ltr  /dev/tty*
````

**Comment by free-x** [comment]
just 2 cent
RPi3 doesn't have uart3/uart5. Only uart0

**Reply by mgrouch** [reply]
Thanks for the correction

**Comment by TonioTunis** [comment]
Dear All
Reading the post, I understand that its possible to get from OpenMarine MacArthur HAT NMEA 2K and NMEA 0183 info
What about the soft power management, the 12v, the IMU info ?
Is this as well integrated and supported by BBN
Thanks in advance

**Reply by jsextl** [reply]
NMEA 2K works perfectly since I installed the device in April this year - sailed about 3000 miles since then and had no issues.

NMEA 0183 is not in use on my boat, but would expect the same.

IMU was not successful for me - the software works, but in my opinion it is hard to find a real good IMU board - I tried three of them - two were very bad - the last one seems to be OK, however I never got the compass calibration to last longer than 1-2 days, so I switched it off again. Maybe that's due to the place I mounted the Raspberry Pi - so others might have more success.

Your question about integration - yes, all software is perfectly integrated into BBN and you just have to configure the things you want to use.

**Reply by mgrouch** [reply]
Are you doing full 360 turn of the boat from time to time to re-calibrate IMU compass?
The Kalman filter used by pypilot is from RTIMULib2 and it's a bit simplified. The model of the Kalman filter doesn't include biases or
estimating biases. So that Kalman filter relies fully on calibration.
I've found more promising Kalman filter https://github.com/thomaspasser/q-mekf and started thinking of implementing just IMU compass with it using m5stack atomS3R.

Thanks

**Reply by TonioTunis** [reply]
Thanks

**Reply by jsextl** [reply]
Yes, I did many 360 turns and this calibrated the compass for a while - but somehow it lost calibration after 1-2 days again and also when heeled the compass did not work very well - and yes, I fixed the Rasperry Pi with a casing and adhesive tape in a very fixed position.

Maybe the problem is, that I have too many other electronics around in the place I installed the Raspberry Pi. Installed an Orca Core for testing in the same position and with the Orca the compass works perfectly after one calibration 2 weeks ago - maybe they are just using a better IMU but that's just speculation.

**Reply by YMita2017** [reply]
"What about the soft power management?" please teach me!

**Reply by YMita2017** [reply]
I tried MacArthur HAT soft power management with BBN OS lysmarine-bbn-full-bookworm_2025-05-18-r1441-raspios-arm64.img.xz clean install, add two line below to /boot/firmware/config.txt 
dtoverlay=gpio-shutdown,gpio_pin=21,active_low=1,gpio_pull=up
dtoverlay=gpio-poweroff,gpiopin=26,active_low=1
But not work.
Also try [https://github.com/bareboat-necessities/lysmarine_gen/issues/389](url) but can't resolve problem.
Please help me!

**Comment by Confidence-DE** [comment]
Hi guys, 

tried to bring a MacArthur HAT to life on a Raspi PI5 with the information in the answer above. CAN interface doesn't show up and socketcan-interface end with an error. I am by far no Linux expert, any chance to point me to help if this happens? Devices connected to the HAT (i2s etc. are working)

Thx for your help. 
Best
Axel


This is what I added to config.txt

#MacArthur Head OVerlays etc
dtoverlay=mcp251xfd,oscillator=20000000,interrupt=25,spi0-1

#MacArthur Head - Dies sollte nur für NMEA0183 relevant sein
dtparam=uart0=on
dtoverlay=uart2-pi5
dtoverlay=uart4-pi5

Ifconfig doesn't show can

socketcan enable error: 
Job for socketcan-interface.service failed because the control process exited with error code.
See "systemctl status socketcan-interface.service" and "journalctl -xeu socketcan-interface.service" for details.


An ExecStart= process belonging to unit socketcan-interface.service has exited.
░░ 
░░ The process' exit code is 'exited' and its exit status is 1.
Jan 08 09:40:54 brain-of-confidence systemd[1]: socketcan-interface.service: Failed with result 'exit-code'.
░░ Subject: Unit failed
░░ Defined-By: systemd
░░ Support: https://www.debian.org/support
░░ 
░░ The unit socketcan-interface.service has entered the 'failed' state with result 'exit-code'.
Jan 08 09:40:54 brain-of-confidence systemd[1]: Failed to start socketcan-interface.service - SocketCAN interface can0 with a baudrate of 250000.
░░ Subject: A start job for unit socketcan-interface.service has failed
░░ Defined-By: systemd
░░ Support: https://www.debian.org/support
░░ 
░░ A start job for unit socketcan-interface.service has finished with a failure.
░░ 
░░ The job identifier is 3976 and the job result is failed.
Jan 08 10:28:38 brain-of-confidence systemd[1]: Starting socketcan-interface.service - SocketCAN interface can0 with a baudrate of 250000...
░░ Subject: A start job for unit socketcan-interface.service has begun execution
░░ Defined-By: systemd
░░ Support: https://www.debian.org/support
░░ 
░░ A start job for unit socketcan-interface.service has begun execution.
░░ 
░░ The job identifier is 4964.
Jan 08 10:28:38 brain-of-confidence ip[4884]: Cannot find device "can0"
Jan 08 10:28:38 brain-of-confidence systemd[1]: socketcan-interface.service: Main process exited, code=exited, status=1/FAILURE
░░ Subject: Unit process exited
░░ Defined-By: systemd
░░ Support: https://www.debian.org/support
░░ 
░░ An ExecStart= process belonging to unit socketcan-interface.service has exited.
░░ 
░░ The process' exit code is 'exited' and its exit status is 1.
Jan 08 10:28:38 brain-of-confidence systemd[1]: socketcan-interface.service: Failed with result 'exit-code'.
░░ Subject: Unit failed
░░ Defined-By: systemd
░░ Support: https://www.debian.org/support
░░ 
░░ The unit socketcan-interface.service has entered the 'failed' state with result 'exit-code'.
Jan 08 10:28:38 brain-of-confidence systemd[1]: Failed to start socketcan-interface.service - SocketCAN interface can0 with a baudrate of 250000.
░░ Subject: A start job for unit socketcan-interface.service has failed
░░ Defined-By: systemd
░░ Support: https://www.debian.org/support
░░ 
░░ A start job for unit socketcan-interface.service has finished with a failure.
░░ 
░░ The job identifier is 4964 and the job result is failed.
~


root@brain-of-confidence:/boot/firmware# lsmod | grep can
can_raw                49152  0
can                    49152  1 can_raw
can_dev                81920  1 mcp251xfd

**Reply by jsextl** [reply]
Hi Axel,
some thoughts about that:

1. Did you configure it in the right place? Since newer versions the config.txt is located in

`/boot/firmware/config.txt`

not directly in /boot/config.txt like in older versions.

2. Maybe the section in config.txt is wrong - I have it configured in the section "all":

```
[all]
dtoverlay=mcp251xfd,spi0-1,oscillator=20000000,interrupt=25
```

3. Can you show the output of `dmesg|grep mcp251xfd`

That should show any messages about the driver loading and should look like the following on a RPi5:

`
root@morningsun:/boot/firmware# dmesg|grep mcp251xfd
[    3.010533] mcp251xfd spi0.1 can0: MCP2518FD rev0.0 (-RX_INT -PLL -MAB_NO_WARN +CRC_REG +CRC_RX +CRC_TX +ECC -HD o:20.00MHz c:20.00MHz m:20.00MHz rs:8.50MHz es:8.33MHz rf:8.50MHz ef:8.33MHz) successfully initialized.
`

Best regards,
Jonas

**Reply by Confidence-DE** [reply]
Hi Jonas, 

1. Yes, this is exactly the place where I edited the config.txt
2. Same here all in the "all" section

dmesg|grep mcp251xfd
[    4.033950] mcp251xfd spi0.1 (unnamed net_device) (uninitialized): Failed to read Oscillator Configuration Register (osc=0x00000000).
[    4.033971] mcp251xfd spi0.1: error -ENODEV: Failed to detect MCP251xFD

This is the output of dmesg

best
Axel

**Reply by jsextl** [reply]
To be honest, I think you have a hardware problem then as the HAT is not detected.

So maybe double-check the connectors, etc. or maybe it is a broken device. Also just try with Rpi5 plus HAT alone without any other connectors, devices, etc. to be sure they don't interfere with each other.
